### PR TITLE
Improve logging of request drain timeout

### PR DESCRIPTION
**Pull request by halter73**
#2353

### CODE REVIEW

File: src/Kestrel.Core/Internal/Http/Http1MessageBody.cs [hunk 1]
```diff
@@ -125,18 +124,49 @@ public override Task StopAsync()
             return _pumpTask;
         }
 
-        protected override async Task OnConsumeAsync()
+        protected override Task OnConsumeAsync()
         {
-            _context.TimeoutControl.SetTimeout(Constants.RequestBodyDrainTimeout.Ticks, TimeoutAction.SendTimeoutResponse);
+            var readTask = _context.RequestBodyPipe.Reader.ReadAsync();
+
+            if (readTask.IsFaulted || readTask.IsCanceled)
```
Comment (davidfowl) [inline src/Kestrel.Core/Internal/Http/Http1MessageBody.cs]: The pipe task is never cancelled.

**Comment by halter73** [comment]
🆙 📅

**Comment by halter73** [comment]
@pakrym @davidfowl One of you please approve this so I don't have to deal with ask mode.

File: src/Kestrel.Core/Internal/Http/Http1MessageBody.cs [hunk 2]
```diff
@@ -125,18 +124,49 @@ public override Task StopAsync()
...
+            var readTask = _context.RequestBodyPipe.Reader.ReadAsync();
+
+            if (readTask.IsFaulted)
+            {
+                return Task.CompletedTask;
+            }
+
+            if (readTask.IsCompletedSuccessfully)
+            {
+                var result = readTask.GetAwaiter().GetResult();
```
Comment (pakrym) [inline src/Kestrel.Core/Internal/Http/Http1MessageBody.cs]: TryRead

**Comment by halter73** [comment]
Here are the logs showing why I cnahged "OnCompletedShouldNotBlockAResponse". It wasn't failing, but it was creating noise.

```
2018-04-13 22:38:05.076 +00:00 [TestLifetime] [Information] Starting test "OnCompletedShouldNotBlockAResponse"
2018-04-13 22:38:05.086 +00:00 [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Debug] Hosting starting
2018-04-13 22:38:05.099 +00:00 [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Debug] Hosting started
2018-04-13 22:38:05.099 +00:00 [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Debug] Loaded hosting startup assembly "Libuv.FunctionalTests"
2018-04-13 22:38:05.100 +00:00 [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id ""0HLD1HSM93P9E"" started.
2018-04-13 22:38:05.100 +00:00 [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Information] Request starting HTTP/1.1 GET http://127.0.0.1:32804/
2018-04-13 22:38:05.111 +00:00 [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id ""0HLD1HSM93P9E"" completed keep alive response.
2018-04-13 22:38:05.117 +00:00 [Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv] [Debug] Connection id ""0HLD1HSM93P9E"" received FIN.
2018-04-13 22:38:05.117 +00:00 [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id ""0HLD1HSM93P9E"" disconnecting.
2018-04-13 22:38:05.117 +00:00 [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Debug] Hosting shutdown
2018-04-13 22:38:05.118 +00:00 [Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv] [Debug] Connection id ""0HLD1HSM93P9E"" sending FIN.
2018-04-13 22:38:11.011 +00:00 [Microsoft.AspNetCore.Server.Kestrel] [Debug] Some connections failed to close gracefully during server shutdown.
2018-04-13 22:38:35.085 +00:00 [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Information] Request finished in 29984.7976ms 200
2018-04-13 22:38:35.088 +00:00 [Microsoft.AspNetCore.Server.Kestrel] [Warning] Connection shutdown abnormally.
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'MemoryPoolBlock'.
   at System.Buffers.ThrowHelper.ThrowObjectDisposedException(ExceptionArgument argument) in /_/.nuget/packages/microsoft.extensions.buffers.sources/2.1.0-preview3-32196/contentFiles/cs/netstandard1.0/Microsoft.Extensions.Buffers.Sources/PipelinesThrowHelper.cs:line 51
   at System.Buffers.MemoryPoolBlock.Dispose() in /_/.nuget/packages/microsoft.extensions.buffers.sources/2.1.0-preview3-32196/contentFiles/cs/netstandard1.0/Microsoft.Extensions.Buffers.Sources/MemoryPoolBlock.Release.cs:line 54
   at System.IO.Pipelines.BufferSegment.ResetMemory()
   at System.IO.Pipelines.Pipe.CompletePipe()
   at System.IO.Pipelines.Pipe.CompleteReader(Exception exception)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequestsAsync[TContext](IHttpApplication`1 application) in /_/src/Kestrel.Core/Internal/Http/HttpProtocol.cs:line 478
2018-04-13 22:38:35.089 +00:00 [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id ""0HLD1HSM93P9E"" stopped.
```

File: src/Kestrel.Core/Internal/Http/Http1MessageBody.cs [hunk 3]
```diff
@@ -125,19 +124,55 @@ public override Task StopAsync()
...
+                {
+                    _context.RequestBodyPipe.Reader.AdvanceTo(readResult.Buffer.End);
+
+                    if (readResult.IsCompleted)
+                    {
+                        return Task.CompletedTask;
+                    }
+                }
+            }
+            catch (BadHttpRequestException ex)
```
Comment (davidfowl) [inline src/Kestrel.Core/Internal/Http/Http1MessageBody.cs]: How would this get thrown from the code above? We're not calling anything that would result in this exception

Comment (davidfowl) [reply src/Kestrel.Core/Internal/Http/Http1MessageBody.cs]: Wait, I liked, hmm it looks like TryRead does throw if the writer completed with an error? That right @pakrym ?

File: src/Kestrel.Core/Internal/Http/HttpProtocol.cs [hunk 4]
```diff
@@ -605,19 +596,13 @@ public void OnHeader(Span<byte> name, Span<byte> value)
...
-                    // (DisposeContext logs StatusCode).
-                    SetBadRequestState(badRequestException);
-                }
-
                 application.DisposeContext(httpContext, _applicationException);
 
-                // StopStreams should be called before the end of the "if (!_requestProcessingStopping)" block
-                // to ensure InitializeStreams has been called.
-                StopStreams();
+                // Even for non-keep-alive requests, try to consume the entire body to avoid RSTs.
```
Comment (davidfowl) [inline src/Kestrel.Core/Internal/Http/HttpProtocol.cs]: So we're consuming the body for non keep alive requests? Didn't we do this before and moved away from it?

**Comment by davidfowl** [comment]
Code is very clean, I have one question I'd like answered before signing off

File: src/Kestrel.Core/Internal/Http/Http1MessageBody.cs (hunk 3 above)
Comment (halter73) [reply src/Kestrel.Core/Internal/Http/Http1MessageBody.cs]: [It does.](https://github.com/dotnet/corefx/blob/0b9a376b08331692ce68ebffb38e3bcf3c05245b/src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs#L760)

File: src/Kestrel.Core/Internal/Http/HttpProtocol.cs (hunk 4 above)
Comment (halter73) [reply src/Kestrel.Core/Internal/Http/HttpProtocol.cs]: Not writing the response prior to consuming the body could lead to hangs with some clients. Not consuming the entire body leads to another problem where some clients (e.g. HttpClient) will treat the request as failed if it receives and RSTs.

Comment (halter73) [reply src/Kestrel.Core/Internal/Http/HttpProtocol.cs]: Anyway, this isn't functionally different from where we called it before. It's still after writing the response, but now it also after disposing the HttpContext so Hosting's request timing logs are now more accurate.

**Pull request merged**